docs/listeners/jsonapi.rst

```diff
@@ -0,0 +1,723 @@
+v1.0 with support for data fetching, data posting, (validation) errors and a ton of
+configurable options to manipulate the generated json allowing you to benefit of instant
+compatibility with JSON API supporting tools and frameworks like Ember Data.
+
+Requirements
+------------
+
+This listener depends on the `neomerx/json-api <https://github.com/neomerx/json-api>`_ 
+composer package which can be installed by running:
+
+.. code-block:: bash
+
+   composer require neomerx/json-api:^0.8.10
+
+Setup
+-----   
+
+Routing
+^^^^^^^
+
+Only controllers explicitely mapped can be exposed as API resources so make sure 
+to configure your global routing scope in ``config/router.php`` similar to:
+
+
+.. code-block:: phpinline
+
+  const API_RESOURCES = [
+    'Countries',
+    'Currencies',
+  ];
+
+  Router::scope('/', function ($routes) {
+    foreach (API_RESOURCES as $apiResource) {
+        $routes->resources($apiResource, [
+            'inflect' => 'dasherize'
+        ]);
+	}
+  }
+
+Controller
+^^^^^^^^^^
+
+Attach the listener using the components array if you want to attach
+it to all controllers, application wide and make sure ``RequestHandler``
+is loaded **before** ``Crud``.
+
+.. code-block:: php
+
+  <?php
+  class AppController extends Controller {
+
+    public function initialize()
+    {
+      $this->loadComponent('RequestHandler');
+      $this->loadComponent('Crud.Crud', [
+        'actions' => [
+          'Crud.Index',
+          'Crud.View'
+        ],
+        'listeners' => ['Crud.JsonApi']
+      ]);
+    }
+
+
+Alternatively, attach the listener to your controllers ``beforeFilter`` 
+if you prefer attaching the listener to specific controllers on the fly.
+
+.. code-block:: php
+
+  <?php
+  class SamplesController extends AppController {
+
+    public function beforeFilter(\Cake\Event\Event $event) {
+      parent::beforeFilter();
+      $this->Crud->addListener('Crud.JsonApi');
+    }
+  }
+
+
+CakeRequest detector
+--------------------
+
+The JsonApi Listener adds the ``is(jsonapi)`` request detector 
+to your ``CakeRequest`` object which check if the request
+contains a ``HTTP Accept`` header  set to ``application/vnd.api+json``.
+
+.. note::
+
+To make sure the listener won't get in your way it will 
+return ``NULL`` for all requests unless ``is('jsonapi')`` is true.
+
+Exception handler
+-----------------
+
+The JsonApi listener overrides the ``Exception.renderer`` for ``jsonapi`` requests,
+so in case of an error, a standardized error in JSON API format will be returned
+for both errors/exceptions and validation errors.
+
+Errors/exceptions
+^^^^^^^^^^^^^^^^^
+
+For standard errors (e.g. 404) and exceptions the listener will
+produce error responses in the following JSON API format:
+
+.. code-block:: json
+
+  {
+    "errors": [
+      {
+        "code": "501",
+        "title": "Not Implemented"
+      }
+    ],
+    "debug": {
+      "class": "Cake\\Network\\Exception\\NotImplementedException",
```

Comment:
Just curious but would this be better as`"class": \Cake\Network\Exception\NotImplementedException::class` ?